src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -534,18 +541,19 @@ InProgressState inProgressState
 
                     async Task<FinalState> BuildFinalStateFromScratchAsync(
                         SolutionCompilationState compilationState,
-                        CompilationTrackerGeneratorInfo generatorInfo,
+                        NoCompilationState noCompilationState,
                         CancellationToken cancellationToken)
                     {
+                        Contract.ThrowIfTrue(noCompilationState.GeneratorInfo.DocumentsAreFinal);
+
                         try
                         {
-                            var compilation = await BuildDeclarationCompilationFromScratchAsync(
-                                generatorInfo, cancellationToken).ConfigureAwait(false);
+                            var allSyntaxTreesParsedState = await BuildAllSyntaxTreesParsedStateFromScratchAsync(
+                                noCompilationState, cancellationToken).ConfigureAwait(false);
 
                             return await FinalizeCompilationAsync(
```

Comment:
finalize compilation became much clearer now as it stated explicitly waht tracker state transition it corresponds to.